dsl_parser/constraints.py

```diff
@@ -218,14 +222,94 @@ class Pattern(Constraint):
     # This class represents a regex constraint.
     # E.g. if self.args = 'abc' then calling `predicate` will only return True
     # when value = "abc".
-    def predicate(self, value):
+    def predicate(self, value, data_type=None, get_method=None):
         if not isinstance(value, text_type):
             raise exceptions.ConstraintException(
                 "Value must be of type string, got type "
                 "'{0}'".format(type(value).__name__))
         return bool(re.match(self.args, value))
 
 
+@register_constraint(name='filter', constraint_data_type=_DICT)
+class Filter(Constraint):
+    def predicate(self, value, data_type=None, get_method=None):
+        if data_type not in ['deployment_id']:
+            raise exceptions.ConstraintException(
+                "Filter constraint is not defined for "
+                "'{0}' data types".format(data_type))
+        if not get_method:
+            raise exceptions.ConstraintException(
+                "Filter constraint requires a get_method "
+                "to retrieve related entities.")
+        try:
+            entities = get_method(value)
+        except Exception:
+            raise exceptions.ConstraintException(
+                "No object referenced by type '{0}' and identified by '{1}' "
+                "has been found".format(data_type, value))
+        if data_type == 'deployment_id':
+            return self.validate_deployment(value, entities)
+
+    def validate_deployment(self, deployment_id, deployments):
+        if len(deployments) < 1:
+            raise exceptions.ConstraintException(
+                "Deployment has not been found: '{0}'".format(deployment_id))
+        for name, definition in self.args.items():
+            if name == 'labels':
+                deployments = self.validate_deployment_labels(
+                    definition, deployments)
+            elif name == 'tenants':
+                deployments = self.validate_deployment_tenants(
+                    definition, deployments)
+            elif name == 'name_pattern':
+                deployments = self.validate_deployment_name_pattern(
+                    definition, deployments)
+            else:
+                raise exceptions.ConstraintException(
+                    "Filter '{0}' is not defined for data type 'deployment_id'"
+                    .format(name))
+        if len(deployments) < 1:
+            raise exceptions.ConstraintException(
+                "No matching deployment has been found for these constraints: "
+                "'{0}'".format(self.args))
+        return True
+
+    @staticmethod
+    def validate_deployment_labels(labels, deployments):
```

Comment:
Can't this and the tenant thing be done as part of the get method by applying filters so you then just check the result count?

Reply:
No, not really.  I don't believe get method supports combining tenants, labels and display_name filters.

Reply:
A func.partial could do it (if I'm thinking of the right thing), and then we should be making the database do the heavy lifting rather than doing it in python I think- not a big issue either way for small environments but could get painful in larger.

Reply:
Unfortunately (?) it's not going to be a simple query to the DB.  Most of the times we'll be using rest-client (see https://github.com/cloudify-cosmo/cloudify-manager/pull/3530/files#diff-a441270002aa034fc667e2efbfa3514f40909d712120a8fa3b3ecd8822193dbcR111)

Reply:
Fair enough, then let's go with this and optimise later if we need to.